Newparticle

alicerobson opened the pull request:
Reissue of Particle3 pull request.

It addresses issues #6 #7 (merging and tidying PFParticle and Particle classes) and #17 which makes use of propagator consistent with Python. 

The simulated particles are now created outside of the simulation and the pdebug outputs are matching with current Python again.

There is also tidying of the use of path.

The code runs under v0.8. I have temporarily left some (commented out) code in PythiaConnecter allowing it to run on earlier version of EDM on Mac if need be.

cbernet commented on papas/simulation/Propagator.h:
```diff
@@ -3,34 +3,35 @@
-   @param[in] detector  Detector through which to propagate
-   */
-  void propagate(const PFParticle& ptc, const Detector& detector) const;
+  virtual void propagateOne(Particle& ptc, const SurfaceCylinder& cyl) const = 0;
```
add line break

alicerobson replied:
done

alicerobson replied:
Do you want line breaks like this everywhere?

cbernet replied:
Hi Alice, there should be a line break between methods, not sure it's the case here, and no linebreak between a docstring and the corresponding method. But no need to change anything else in this PR, you can always do that later.

cbernet commented on papas/simulation/Propagator.h:
```diff
@@ -3,34 +3,35 @@
+    @param[in] ptc particle that is to be propagated
+    @param[in] detector  Detector through which to propagate
+    */
+
```
rm line break

cbernet commented on papas/simulation/src/Simulator.cpp:
```diff
@@ -25,39 +25,46 @@ Simulator::Simulator(const Event& papasevent, const ListParticles& particles, co
-    simulateParticle(p);
+  //make sure we can process the particles in order if needed (highest energy first)
+  Ids ids;
+  for (auto& p : particles)
```
const auto&

alicerobson replied:
Thanks

cbernet commented on papas/simulation/src/Simulator.cpp:
```diff
@@ -87,14 +97,12 @@ void Simulator::simulateHadron(PFParticle& ptc) {
       /// ecal path length can be infinite in case the ecal
       /// has lambda_I = 0 (fully transparent to hadrons)
-      auto path = ptc.path();
+      auto path = std::static_pointer_cast<Helix>(ptc.path());
```
Why is this cast needed?

alicerobson replied:
Good spot. Its not needed in this instance. Removed.

cbernet commented on papas/simulation/src/Simulator.cpp:
```diff
@@ -218,13 +204,13 @@ PFParticle& Simulator::addGunParticle(int pdgid, double charge, double thetamin,
-  if (ptc.hasNamedPoint(papas::Position::kEcalIn)) {
-    TVector3 pos = ptc.pathPosition(papas::Position::kEcalIn);
+  if (ptc.path()->hasNamedPoint(papas::Position::kEcalIn)) {
+    TVector3 pos = ptc.path()->namedPoint(papas::Position::kEcalIn);
```
why not a const TVector3& ?

alicerobson replied:
Indeed.

cbernet commented on papas/simulation/src/Simulator.cpp:
```diff
@@ -238,17 +224,17 @@ Cluster Simulator::makeAndStoreEcalCluster(const PFParticle& ptc, double fractio
-  if (ptc.hasNamedPoint(papas::Position::kHcalIn)) {
-    TVector3 pos = ptc.pathPosition(papas::Position::kHcalIn);
+  if (ptc.path()->hasNamedPoint(papas::Position::kHcalIn)) {
+    TVector3 pos = ptc.path()->namedPoint(papas::Position::kHcalIn);
```
same, please check all uses of namedPoint

alicerobson replied:
Done

cbernet commented on tests/unittest.cpp:
```diff
@@ -104,14 +104,14 @@ TEST_CASE("Helixpath") {  /// Helix path test
   helixprop.propagateOne(particle, cyl1);
-  auto tvec = particle.pathPosition(cyl1.layer());
-  auto particle2 = PFParticle(211, -1, TLorentzVector{0., 2, 1, 5}, 2, 'r', TVector3{0, 0, 0}, 3.8);
+  auto tvec = particle.path()->namedPoint(cyl1.layer());
```
const auto&

alicerobson replied:
Done

cbernet commented on tests/unittest.cpp:
```diff
@@ -104,14 +104,14 @@ TEST_CASE("Helixpath") {  /// Helix path test
+  auto particle2 = Particle(211, -1, TLorentzVector{0., 2, 1, 5}, 2, 'r', TVector3{0, 0, 0}, 3.8);
   helixprop.propagateOne(particle2, cyl1);
-  auto tvec2 = particle2.pathPosition(cyl1.layer());
+  auto tvec2 = particle2.path()->namedPoint(cyl1.layer());
```
same

alicerobson replied:
Done

alicerobson commented on examples/PythiaConnector.cpp:
```diff
@@ -46,72 +46,48 @@ PythiaConnector::PythiaConnector(const char* fname) : m_store(podio::EventStore(
 }
 
-papas::ListParticles PythiaConnector::makePapasParticlesFromGeneratedParticles(const fcc::MCParticleCollection* ptcs) {
+void PythiaConnector::makePapasParticlesFromGeneratedParticles(const fcc::MCParticleCollection* ptcs,
```
Hi Colin
I completely agree that it could be simpler. However, if I am to match the pdebug outputs from python, I must first sort the generated particles and then create the papas particles from the sorted gen particles. Otherwise the identifiers of the particles differ between papas and cpp. 
Would it be better to change the way it works in Python (and implement your comments above)?

cbernet replied:
For now let's leave it like this. 
At some point, we should change that on both the c++ and python sides. can you create an issue with low priority to keep track of that?

alicerobson commented:
Thanks Colin!
I have addressed all comments except for the one concerning makePapasParticlesFromGeneratedParticles which I guess we should discuss.
Alice

cbernet commented:
@alicerobson actually I don't see your latest changes, can you update the branch?

alicerobson commented:
Sure
